### DISCUSSION TITLE
Input box or dropdown box as the filter in chat

### DISCUSSION THREAD

**Discussion by Phoebecdy91** [zowe/zowe-chat#67] [General]
When chat cannot response them with the data they request, what could be a better user experience?
- let users to type sentence again?
- let users to take quick action in some input box or dropdown menu?

<img width="604" alt="image" src="https://user-images.githubusercontent.com/23204849/178736846-5c1724b7-656f-4e37-9524-ca9771b611bd.png">

**Comment by Phoebecdy91** [comment]
From my opinion:
- This is happening when user type something wrong and cannot get the job he/she expect.
- we don't want the user to type long sentence of the cmd again, especially now we don't support NLP. And it could **not be easy for all the users to remember** the syntax. It is easy to make mistake.
- It is the immediate response when user didn't get what they want. And what users **expecting now is how I can get it.**
- Quick action options can s**ave some time and effort** for users.
- There might be a lot of users in this channel and see this msg. Maybe more than one person will try to input it and request info. That might be messy. But it is similar as they type sentence to request. More than 1 user will using cmd to request data.

Any other thoughts are welcome.

**Comment by Wusong-Fang** [comment]
Hi Phoebe, 

Thanks a lot for creating the issue and bring up the discussion. I don't think it's a good idea to add some filters like that. More detailed input and comments from me can be seen below:

1. First of all, it would be better to make clear and understand what kind of major user experience should be provided by Zowe Chat as one product. Chatting or mouse navigation? If Zowe Chat is full of mouse navigation, then we'd better ask ourselves whether we still need to create a new product and call it as Zowe Chat. Zowe Desktop already is there and has a great web UI for users to get all same data of z/OS resources and manage them there via fully mouse navigation.

2. We must know the major different of user behavior and experience between chatting and web UI navigation. From my standpoint of view, the major operation for chatting is **typing** or **talking**, not mouse clicking. Of cause we can add some interactive components to help users to drill down with a cherry on top. But the least the best. It does not mean that we can add interactive components everywhere just like what we usually do in web UI.

3. As to your question `what kind of response must be provided when user can't get for his input`, I would suggest we'd better consider and think from the product level. We can not do it case by case. We all know that the consistency is one of major requirements for product feature design and also is very important for users to learn a new product. I'd like to know what are our high level consideration and design for that when we operate different z/OS resources from designers' perspective. Will we plan to provide input box style response for all errors?

4. As to your concern `type long sentence of the cmd again`, we must think:  
   1. How often will users use such kind of long sentence or command? And when? 
   2. Is our workflow designed correctly to cause users to use such kind of long sentence or command so frequently? 
   3. Is there any way to simplify the input? Per what I know, abbreviation already exists in Zowe CLI commands as one option. Copy & paste works too. IBM Z ChatOps product also has several different great ideas to solve such kind of known issue. 

6. As to the immediate response, I don't think filtering with `input box` is an immediate one. In my eyes, such kind of response is indirect and look like very strange since users want to get A, but the response provides B. Moreover more actions (typing + clicking) have to be taken by users if using filters like that.

7. Same feature usually is supported by different chat tools in different way. Moreover there are lots of different limitations brought about by different chat tools. We can't break such kind of limitations to implement all mouse navigation features like we usually do in web UI.

8. As to the sample input `@Zowe Chat, zowe zos-jobs view job-status-by-job07484`, I think there are some misunderstanding in your design: 
   1. there is no way or every difficult to extract the keyword `job07484` from a word `job-status-by-job07484` as job id. There are tons of possible results for a word inputted by users if we really analyze it. Of cause it's not a problem to analyze a specific word case by case, but we must consider whether we support the word level analysis from the whole product level. 
   2. There is no way to submit the inputted value of input box without a button in chat tools per my understanding. That's another reason why filters are not an immediate response. More actions (typing + clicking) must be taken by users. 
   3. The comma `,` is redundant if we assume what users' input is a command.
   4. Wildcard is supported in the filter design. How should we do if multiples jobs are returned? Especially when the target job is not included there? Please be noted that the current user case is to query details of a specific job.

10. Last but not least, NLU (Natural Language Understanding) will be supported by Zowe Chat in the long run (maybe our next release).

**Comment by Phoebecdy91** [comment]
Hi Wusong, this is a very valuable discussion. Thanks for putting up your point of view. I agree for some points that we need to consider Zowe Chat as one product and how its user experience will be.

And First of all, I would love to claim the key point of Zowe Chat: it is the 4th way (apart from CLI, APIML and Explorer) for helping user  get info from mainframe, using chat tool. **And our intention is helping user get the info more naturally, conveniently and efficiently.** So conversational interaction is good way, but it is not the only way. When you are using Slack, Teams..., it is not like you are using terminal that keyboard is the only way. 

And what I would like to emphasize here is that, the proposal I put up here is **not blocking user from using chat but give them a second choice, which is a quick filter if they need.** It is just an intelligent assistant that like the bot can detect what user needs and provide a short cut. It is not happening anywhere. You can see from our other design scenario, chatting is still a main way to have interaction. 

I am so happy to know that we gonna support NLP in the near future. Really great! If our NLP is well trained and continuous conversation is clever enough, I also agree with you, that we don't need this kind of WebUI thing as the assistant. But if in MVP, we use CLI command as the one for user to input and we cannot supplement all the sentences that users will use to request data, in case we are going to show too many "sorry" and ask them to do the repeated work, it would be better for us to think about how to save their effort.

**Again, it is just we know what user will do so we provide the short cut to provide convenience.** And this is how we can make our Chat look more and more "intelligent". And I agree with the development effort that may be it needs to be implemented case to case. At least we can improve the user experience, starting from some cases that we can foresee.

Last but not least, **one of the power of Chat is the mobile end.** The user experience is quite different. Key board and info displaying is really limited. So long sentence is not a smart choice.

I just refine the design with another option. When we cannot find the info or user type it wrong. We can guide user to know what correct syntax will be. What's more, provide another btn "Quick Filter". 
If user prefers typing, he/she can continue with manual input cmd. If user were the one who doesn't know syntax quite well, Quick Filter would be a better choice. He/She just need to click the btn, then the input box can be shown. 
<img width="693" alt="image" src="https://user-images.githubusercontent.com/23204849/179436612-01213ef1-9190-4902-9ff0-71836d27db88.png">
<img width="694" alt="image" src="https://user-images.githubusercontent.com/23204849/179436621-56479aee-694f-4374-b402-4b3793a35c10.png">
